Simpler count store implementation.

maxsumrall opened the pull request:
A count store which provides methods for updates, reads, and snapshots. Requesting a snapshot will return a snapshot of the counts store at a transaction ID equal to or greater than the requested transaction ID. Snapshots also provide methods for (de)serializing to a log.

Commit:
Simpler count store implementation.

A count store which provides methods for updates, reads, and snapshots. Requesting a snapshot will return a snapshot of the counts store at a transaction ID equal to or greater than the requested transaction ID. Snapshots also provide methods for (de)serializing to a log.

thobe commented:
How does this implementation ensure consistent snapshots given transactions being applied out-of-order?

thobe commented:
This implies a new (incompatible) store format for the counts, right?

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/CountStore.java:
```diff
@@ -0,0 +1,127 @@
+        {
+            pairs.forEach(
+                    ( key, value ) -> map.compute( key, ( k, v ) -> v == null ? value : updateEachValue( v, value ) ) );
+            lastTxId.offer( txId, new long[]{1} );
```
Let's have a static final field for `new long[]{1}`

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/CountStore.java:
```diff
@@ -0,0 +1,127 @@
+
+    }
+
+    public static ConcurrentHashMap<CountsKey,long[]> copyOfMap( ConcurrentHashMap<CountsKey,long[]> nextMap )
```
private?

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/CountStore.java:
```diff
@@ -0,0 +1,127 @@
+        try
+        {
+            validatedSnapshot = snapshot;
+            snapshot = null;
```
you can move `snapshopt = null` the finally-block and simply `return snapshot` from the try-block and remove `validatedSnapshot`

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/CountStore.java:
```diff
@@ -0,0 +1,127 @@
+import org.neo4j.kernel.impl.store.counts.keys.CountsKey;
+import org.neo4j.kernel.impl.util.ArrayQueueOutOfOrderSequence;
+
+public class CountStore
```
I think we need an interface for the counts store.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/CountStore.java:
```diff
@@ -0,0 +1,127 @@
+        return v;
+    }
+
+    public Snapshot snapshot( long txId )
```
I think we need to document that the snapshot is thread safe wrt updates but you cannot take 2 snapshots from 2 different thread in parallel.

davidegrohmann replied:
Perhaps we can add `assert snapshot == null` as well

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/Snapshot.java:
```diff
@@ -0,0 +1,163 @@
+import static org.neo4j.kernel.impl.store.counts.keys.CountsKeyType.INDEX_STATISTICS;
+import static org.neo4j.kernel.impl.store.counts.keys.CountsKeyType.value;
+
+public class Snapshot
```
Maybe `CountsSnapshot` is a better name?

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/Snapshot.java:
```diff
@@ -0,0 +1,163 @@
+
+public class Snapshot
+{
+    private ConcurrentHashMap<CountsKey,long[]> map;
```
Since a snapshot is "immutable" we can have a regular `Map` here.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/Snapshot.java:
```diff
@@ -0,0 +1,163 @@
+        return txId;
+    }
+
+    public static void serialize( WritableLogChannel channel, Snapshot snapshot ) throws UnknownKey, IOException
```
I would like this class to be a simple data class. So I would suggest to move `serialize` to a different class maybe named `CountsSerializer`.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/Snapshot.java:
```diff
@@ -0,0 +1,163 @@
+        }
+    }
+
+    public static Snapshot deserialize( ReadableLogChannel channel ) throws UnknownKey, IOException
```
Same as above for `deserialize`

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/keys/CountsKeyType.java:
```diff
@@ -19,9 +19,11 @@
+
 public enum CountsKeyType
 {
-    EMPTY( 0 ), ENTITY_NODE( 2 ), ENTITY_RELATIONSHIP( 3 ), INDEX_STATISTICS( 4 ), INDEX_SAMPLE( 5 );
```
Let's postpone the removal of `EMPTY` since I guess  it is still use by the old code.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/CountStore.java:
```diff
@@ -0,0 +1,127 @@
+{
+    private final ReadWriteLock lock;
+    private final ConcurrentHashMap<CountsKey,long[]> map;
+    private volatile ArrayQueueOutOfOrderSequence lastTxId;
```
This should be `final` and not `volatile`.  Moreover the type here should be `OutOfOrderSequence`.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/Snapshot.java:
```diff
@@ -0,0 +1,163 @@
+        long txid = channel.getLong();
+        int size = channel.getInt();
+
+        ConcurrentHashMap<CountsKey,long[]> map = new ConcurrentHashMap<>( size );
```
This can be a simple `HashMap`

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/util/ArrayQueueOutOfOrderSequence.java:
```diff
@@ -33,6 +33,7 @@
     private long[] highestGapFreeMeta;
     private final SequenceArray outOfOrderQueue;
     private long[] metaArray;
+    private long highestEverSeen = 0;
```
`volatile`

davidegrohmann replied:
You should also initialize `highestEverSeen` in `set( long number, long[] meta )` and perhaps print it in `toString`

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/util/ArrayQueueOutOfOrderSequence.java:
```diff
@@ -44,6 +45,7 @@ public ArrayQueueOutOfOrderSequence( long startingNumber, int initialArraySize,
     @Override
     public synchronized boolean offer( long number, long[] meta )
     {
+        highestEverSeen = Math.max(highestEverSeen, number);
```
code style here is off

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/countStore/CountStoreIntegrationTest.java:
```diff
@@ -0,0 +1,198 @@
+        workerExecutorService.awaitTermination( 100, TimeUnit.SECONDS );
+        snapshotExecutorService.awaitTermination( 100, TimeUnit.SECONDS );
+
+
```
remove empty lines

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/countStore/CountStoreIntegrationTest.java:
```diff
@@ -0,0 +1,198 @@
+        int numberOfUpdates = 900;
+        CountStore countStore = new CountStore();
+        IntermediateStateTestManager intermediateStateTestManager = new IntermediateStateTestManager( numberOfUpdates );
+        ExecutorService workerExecutorService = Executors.newFixedThreadPool( 10 );
```
we can use a `ThreadingRule` instead to make this test more readable.

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/countStore/CountStoreSnapshotSerializerIntegrationTest.java:
```diff
@@ -0,0 +1,132 @@
+    {
+        //GIVEN
+        FileSystemAbstraction fs = new DefaultFileSystemAbstraction();
+        File tempFile = new File( "temp.dat" );
```
Use `File.createTempFile( ... )` instead

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/countStore/CountStoreMapGenerator.java:
```diff
@@ -0,0 +1,74 @@
+ */
+public class CountStoreMapGenerator
+{
+    public static ConcurrentHashMap<CountsKey,long[]> simpleCountStoreMap( int num )
```
Why `ConcurrentHashMap` and no simply `Map`?

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/countStore/CountStoreSnapshotSerializerIntegrationTest.java:
```diff
@@ -0,0 +1,132 @@
+/**
+ * Serializes and deserialize count stores to test that they produce the same count stores.
+ */
+public class CountStoreSnapshotSerializerIntegrationTest
```
We can drop the `Integration` part.

davidegrohmann replied:
Nevermind.

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/countStore/CountStoreSnapshotSerializerTest.java:
```diff
@@ -0,0 +1,373 @@
+import static org.neo4j.kernel.impl.store.counts.keys.CountsKeyType.INDEX_SAMPLE;
+import static org.neo4j.kernel.impl.store.counts.keys.CountsKeyType.INDEX_STATISTICS;
+
+public class CountStoreSnapshotSerializerTest
```
Can we split this test in 2 parts: one for serialization and one for deserialization?

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/countStore/CountStoreTest.java:
```diff
@@ -0,0 +1,154 @@
+        CountStore countStore = new CountStore();
+        Map<CountsKey,long[]> update = new HashMap<>();
+        NodeKey key = CountsKeyFactory.nodeKey( 1 );
+        update.put( key, new long[]{1} );
```
`new long[]{1}` should be a static final field

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/countStore/DiffMapValuesTest.java:
```diff
@@ -0,0 +1,89 @@
+        /**
+         * Works because the JVM uses the same object for Long objects if they're < ~100
+         */
+        Assert.assertSame( 5L, map.get( nodeKey( 1 ) )[0] );
```
Why `assertSame` and not `assertEquals`?

davidegrohmann replied:
Actually we should use `assertArrayEquals`

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/countStore/IntermediateStateTestManagerTest.java:
```diff
@@ -0,0 +1,92 @@
+        return countStoreMapsAreEqual( snapshotA.getMap(), snapshotB.getMap() );
+    }
+
+    public static boolean countStoreMapsAreEqual( Map<CountsKey,long[]> mapA, Map<CountsKey,long[]> mapB )
```
This can be implemented as a `org.hamcrest.Matcher`

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/log/InMemoryLogChannel.java:
```diff
@@ -35,9 +35,22 @@ public void flush() throws IOException
+
+    public InMemoryLogChannel()
+    {
+        this(new byte[1000]);
```
code style off

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/log/InMemoryLogChannel.java:
```diff
@@ -35,9 +35,22 @@ public void flush() throws IOException
+        this(new byte[1000]);
+    }
+
+    public InMemoryLogChannel(byte[] bytes)
```
code style off

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/CountStore.java:
```diff
@@ -0,0 +1,127 @@
+
+    public static ConcurrentHashMap<CountsKey,long[]> copyOfMap( ConcurrentHashMap<CountsKey,long[]> nextMap )
+    {
+        ConcurrentHashMap<CountsKey,long[]> newMap = new ConcurrentHashMap<>();
```
Probably we can construct a simple `HashMap` here

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/Snapshot.java:
```diff
@@ -0,0 +1,163 @@
+                NodeKey nodeKey = (NodeKey) key;
+                channel.put( ENTITY_NODE.code );
+                channel.putInt( nodeKey.getLabelId() );
+                channel.putLong( value[0] );
```
we can add `assert value.length == 1` here and corresponding assertions for the other cases as well

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/util/ArrayQueueOutOfOrderSequence.java:
```diff
@@ -57,6 +59,11 @@ public synchronized boolean offer( long number, long[] meta )
+    {
+        return this.highestEverSeen;
+    }
+
```
We should add tests around this.

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/countStore/DiffMapValuesTest.java:
```diff
@@ -0,0 +1,89 @@
+/**
+ * A test for the test class generating all intermediate states the count
+ */
+public class DiffMapValuesTest
```
I don't think we should testing test code. Let's simply move the static methods where they are used.

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/countStore/IntermediateStateTestManagerTest.java:
```diff
@@ -0,0 +1,92 @@
+/**
+ * A test for the test class generating all intermediate states the count
+ */
+public class IntermediateStateTestManagerTest
```
We should not be testing test code. Let's drop this class and move needed code somewhere else.

davidegrohmann commented:
Let's start taking care of my comments above. After that we can do a second review.

Commit:
Improved code quality for simpler counts store.

Improved code style.
Moved serialization and deserialization to their own classes, making CountsSnapshot a data class only.
Removed tests of tests which were redundant.
Added test for the highestEverSeen method in OutOfOrderSequence.
Fixed bug for highestEverSeen in OutOfOrderSequence.
Removed the writelock in CountsStore during the final part of getting a snapshot. A reordering of statements in UpdateAll makes it redundant while still preserving thread safety in updateAll().
 Explicitly throw exception when attempting to get a snapshot while previous snapshot attempt is underway.

maxsumrall commented:
@thobe Consistent snapshots are achieved by returning either the requested snapshot txid or a snapshot with a txid of the highest yet applied transaction (not the highest_gap_free_number), whichever is the larger value. Once the snapshot is made, we wait and let the transactions with a lower txid be applied to both the store and our snapshot until the snapshot is consistent for its txid. Incoming transactions with a txid higher than what the snapshot is waiting for are only applied to the store.

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/CountsSnapshot.java:
```diff
@@ -0,0 +1,46 @@
+
+public class CountsSnapshot
+{
+    private Map<CountsKey,long[]> map;
```
Fields can be final

maxsumrall commented:
@thobe Yes, this is incompatible with the current store formats for the counts.

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/CountsSnapshotSerializer.java:
```diff
@@ -0,0 +1,95 @@
+            {
+
+            case ENTITY_NODE:
+                assert value.length == 1;
```
Can we please use regular `if` statements instead of asserts?

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/CountsSnapshotDeserializer.java:
```diff
@@ -0,0 +1,83 @@
+                value = new long[]{channel.getLong(), channel.getLong()};
+                map.put( key, value );
+                break;
+            }
```
Would be nice to have `default` case that throws exception

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/CountsSnapshotSerializer.java:
```diff
@@ -0,0 +1,95 @@
+                channel.putLong( value[0] );
+                channel.putLong( value[1] );
+                break;
+
```
Would be nice to have `default` case that throws exception

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/CountsStore.java:
```diff
@@ -0,0 +1,48 @@
+ */
+public interface CountsStore
+{
+    long[] get( CountsKey key );
```
Would be nice to have javadoc here describing what is the returned `long[]`

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/InMemoryCountsStore.java:
```diff
@@ -0,0 +1,146 @@
+        try
+        {
+            Predicates
+                    .awaitForever( () -> lastTxId.getHighestGapFreeNumber() >= snapshot.getTxId(), 100, MILLISECONDS );
```
We've had problems with `awaitForever` type statements in the current counts store. Await should be bounded.

tinwelint replied:
It doesn't make sense to have a timer, it _should_ be indefinite wait. There's no case for a transaction never completing. Either it completes successfully or fails.

lutovich replied:
I can't agree. Indefinite wait means that database will never shutdown and user will kill -9 it. This is an abnormal situation but we should not hang forever in any case.

davidegrohmann replied:
I think it should be an indefinite wait, but it should also check periodically the database health and fail if it is not healthy. That should be enough to solve the hanging on shutdown and make sure we don't cause false timeouts.

lutovich replied:
Sounds good to me. Since this counts store is not hooked in yet it is than ok to keep code as is with a `todo` added, I think

tinwelint replied:
Yes, agreed. That's the best approach @davidegrohmann

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/InMemoryCountsStore.java:
```diff
@@ -0,0 +1,146 @@
+        }
+        catch ( InterruptedException ex )
+        {
+            throw Exceptions
```
Would be good to restore the interruption status of this thread with `Thread.currentThread().interrupt();`

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/keys/CountsKeyType.java:
```diff
@@ -19,14 +19,34 @@
+        case 5:
+            return CountsKeyType.INDEX_SAMPLE;
+        default:
+            throw new UnknownKey("Parsed key type from count store deserialization of unknown type.");
```
Can we throw unchecked exception here?

lutovich commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/countStore/InMemoryCountsStoreCountsSnapshotSerializerIntegrationTest.java:
```diff
@@ -0,0 +1,130 @@
+        {
+            long[] value = recovered.getMap().get( pair.getKey() );
+            Assert.assertNotNull( value );
+            Assert.assertTrue( Arrays.equals( value, pair.getValue() ) );
```
`Assert#assertArrayEquals()`

lutovich commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/countStore/IntermediateStateTestManager.java:
```diff
@@ -0,0 +1,148 @@
+{
+    private int numberOfUpdates;
+    private int id = 1;
+    private ConcurrentHashMap[] maps;
```
Raw types

lutovich commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/InMemoryCountsStore.java:
```diff
@@ -0,0 +1,146 @@
+    @Override
+    public CountsSnapshot snapshot( long txId )
+    {
+        if ( snapshot != null )
```
Reading `snapshot` is racy here. Not sure if/how we need to fix this.

chrisvest commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/CountsStore.java:
```diff
@@ -0,0 +1,48 @@
+{
+    long[] get( CountsKey key );
+
+    void updateAll( long txId, Map<CountsKey,long[]> updates );
```
Javadocs on `updateAll` would also be nice. Generally, interfaces and their methods should always have javadocs.

Commit:
Improvements to CountsStore implementation.

Slightly additional testing and some refactoring based on the received feedback. Some tests changed, revealing a potential bug which has been fixed.

Commit:
Improvements to CountsStore implementation.

Change locking order in snapshot method cleanup test assertions, and removed throwing UnknownKey in (de)serialize.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/InMemoryCountsStore.java:
```diff
@@ -0,0 +1,170 @@
+        if ( snapshot != null )
+        {
+            throw new IllegalStateException( "Cannot perform snapshot while another snapshot is processing." );
+        }
```
This has to be move in the `try` block of the `try-finally` otherwise on failure you'll never release the lock.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/CountsSnapshotDeserializer.java:
```diff
@@ -0,0 +1,88 @@
+
+public class CountsSnapshotDeserializer
+{
+    public static CountsSnapshot deserialize( ReadableLogChannel channel ) throws IOException, UnknownKey
```
Where is `UnknownKey` thrown?

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/countStore/CountsSnapshotSerializer.java:
```diff
@@ -0,0 +1,116 @@
+public class CountsSnapshotSerializer
+{
+    public static void serialize( WritableLogChannel channel, CountsSnapshot countsSnapshot )
+            throws UnknownKey, IOException
```
Same comment here for `UnknownKey`.

davidegrohmann commented on community/kernel/src/main/java/org/neo4j/kernel/impl/store/counts/keys/CountsKeyType.java:
```diff
@@ -19,14 +19,34 @@
         this.code = (byte) code;
     }
+
+    public static CountsKeyType value(byte val) throws UnknownKey
```
Here again `UnknownKey`.

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/countStore/InMemoryCountsStoreCountsSnapshotSerializerIntegrationTest.java:
```diff
@@ -0,0 +1,129 @@
+public class InMemoryCountsStoreCountsSnapshotSerializerIntegrationTest
+{
+    @Test
+    public void smallWorkloadOnInMemoryLogTest() throws IOException, UnknownKey
```
Again `UnknownKey`. Perhaps here the best is just allow the tests to throw `Exception`

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/countStore/InMemoryCountsStoreCountsSnapshotSerializerTest.java:
```diff
@@ -0,0 +1,285 @@
+    ByteBuffer serializedBytes;
+
+    @Before
+    public void setup() throws IOException, UnknownKey
```
Same here, simply allow tests to throw `Exception`

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/countStore/InMemoryCountsStoreSnapshotDeserializerTest.java:
```diff
@@ -0,0 +1,221 @@
+    ByteBuffer serializedBytes;
+
+    @Before
+    public void setup() throws IOException, UnknownKey
```
One more here

davidegrohmann commented on community/kernel/src/test/java/org/neo4j/kernel/impl/store/countStore/InMemoryCountsStoreTest.java:
```diff
@@ -0,0 +1,153 @@
+        countStore.updateAll( 1, update );
+
+        //THEN
+        Assert.assertNull( countStore.get( null ) );
```
why `assertNull`? shouldn't this throw a NPE?
